lib/plugins/aws/package/compile/events/cloudWatchLog/index.js

```diff
@@ -15,133 +15,206 @@ class AwsCompileCloudWatchLogEvents {
-              "Type": "AWS::Lambda::Permission",
-              "Properties": {
-                "FunctionName": { "Fn::GetAtt": ["${
-              lambdaLogicalId}", "Arn"] },
-                "Action": "lambda:InvokeFunction",
-                "Principal": {
-                  "Fn::Join": [ "", [
-                  "logs.",
-                  { "Ref": "AWS::Region" },
-                  ".",
-                  { "Ref": "AWS::URLSuffix" }
-                  ] ]
-                },
-                "SourceArn": {
-                  "Fn::Join": [ "", [
-                  "arn:",
-                  { "Ref": "AWS::Partition" },
-                  ":logs:",
-                  { "Ref": "AWS::Region" },
-                  ":",
-                  { "Ref": "AWS::AccountId" },
-                  ":log-group:",
-                  "${LogGroupName}",
-                  ":*"
-                  ] ]
-                }
-              }
-            }
-          `;
+        const newCloudWatchLogRuleObject = {
+          [cloudWatchLogLogicalId]: JSON.parse(cloudWatchLogRuleTemplate),
+        };
 
-            const newCloudWatchLogRuleObject = {
-              [cloudWatchLogLogicalId]: JSON.parse(cloudWatchLogRuleTemplate),
-            };
+        const newPermissionObject = {
+          [lambdaPermissionLogicalId]: JSON.parse(permissionTemplate),
+        };
 
-            const newPermissionObject = {
-              [lambdaPermissionLogicalId]: JSON.parse(permissionTemplate),
-            };
+        _.merge(this.serverless.service.provider.compiledCloudFormationTemplate.Resources,
+          newCloudWatchLogRuleObject, newPermissionObject);
 
-            _.merge(this.serverless.service.provider.compiledCloudFormationTemplate.Resources,
-              newCloudWatchLogRuleObject, newPermissionObject);
-          }
+        // return a new promise that will check the resource limit exceeded against the function
+        return this.checkAndFixLogGroupSubscriptionFilterResourceLimitExceeded({
+          logGroupName: LogGroupName,
+          functionName,
         });
-      }
+      });
+
+      return Promise.all(eventPromises);
+    });
+
+    return Promise.all(promises);
+  }
+
+  /**
+   * @description Cloudwatch imposes a hard limit of 1 subscription filter per log group.
+   * If we change a cloudwatchLog event entry to add a subscription filter to a log group
+   * that already had one before, it will throw an error because CloudFormation firstly
+   * tries to create and replace the new subscription filter (therefore hitting the limit)
+   * before deleting the old one. This precompile process aims to delete existent
+   * subscription filters of functions that a new filter was provided, by checking the
+   * current ARN with the new one that will be generated.
+   * See: https://git.io/fpKCM
+  */
+  checkAndFixLogGroupSubscriptionFilterResourceLimitExceeded({ logGroupName, functionName }) {
+    return new Promise((resolve, reject) => {
```

Comment:
Could you please rewrite this with out a `new Promise`? Namley, `cloudWatchLogs.describeSubscriptionFilters({ logGroupName }).promise()` is already returning a promise, so there's no reason to wrap that in `new Promise((resolve, reject) => {/*stuff*/})`

Reply:
I can't just return this promise directly because the flow may not even reach to there when:

- the describeSubscriptionFilters call fails (log group doesn't exist yet)
- getAccountId call fails

Reply:
Still shouldn't be an issue (note, I've also modified it to get account & subscription filters in parallel):
```javascript
    const region = this.provider.getRegion();
    const serviceName = this.serverless.service.getServiceName();
    const stage = this.provider.getStage();
    const cloudWatchLogs = new this.provider.sdk.CloudWatchLogs({ region });
    return Promise.all([
        cloudWatchLogs.describeSubscriptionFilters({ logGroupName }).promise(),
        this.provider.getAccountId()])
      .then(([{ subscriptionFilters: [subscriptionFilter] }, accountId]) => {
        // log group doesn't have any subscription filters currently
        if (!subscriptionFilter) {
          return;
        }
        const { destinationArn: oldDestinationArn, filterName } = subscriptionFilter;
        // eslint-disable-next-line max-len
        const newDestinationArn = `arn:aws:lambda:${region}:${accountId}:function:${serviceName}-${stage}-${functionName}`;
            // everything is fine, just resolve
        if (oldDestinationArn === newDestinationArn) {
            return;
        }
        /*
        if the destinations functions' arns doesn't matter, we need to delete the current
        subscription filter to prevent the resource limit exceeded error to happen
        */
        return cloudWatchLogs.deleteSubscriptionFilter({ logGroupName, filterName }).promise();
      })
      /*
        it will throw when trying to get subscription filters of a log group that was just added
        to the serverless.yml (therefore not created in AWS yet), we can safely ignore this error
        and just resolve
      */
      .catch(() => null);
```

Reply:
Ah, you're right! I'll change. I wish we had support to `async/await` to be honest, but we need to have support for NodeJS 4/6 yet here, so it shouldn't be available soon... anyways, I'll make that change.

I'm also thinking right now that it is not necessary to get the accountId every time. I can get it outside and pass as a parameter, I'll rework on this promise flow entirely.